state/src/skip_list.rs

```diff
@@ -0,0 +1,771 @@
+        Op::Delete => delete_item(*account, element, trie),
+        Op::Modify(_) => Ok(()),
+    }
+}
+
+pub fn delete_item<T, D>(
+    account: AccountAddress,
+    element: &T,
+    trie: &mut EthTrie<D>,
+) -> Result<(), TrieError>
+where
+    T: Clone + Listable + Ord + Serialize + DeserializeOwned + 'static,
+    D: DB,
+{
+    let bottom_key = SkipListKey::new(account, 0, element);
+
+    // The key is not present, so there is nothing to do.
+    if bottom_key.trie_value(trie)?.is_none() {
+        return Ok(());
+    }
+
+    let head_key = SkipListHeadKey::<T>::new(account);
+    let head = head_key.trie_value(trie)?;
+
+    let Some(first_value) = head.first_value else {
+        // The list is empty, so there is nothing to delete.
+        return Ok(());
+    };
+
+    let prevs = search_item(account, head.max_levels, first_value, element, trie)?;
+
+    // Search returns keys that are strictly less than the search element.
+    // If the search element is in the list then it will be the value of the returned key
+    for (key, value) in prevs.into_iter().flatten() {
+        if value.next_value.as_deref() != Some(element) {
+            continue;
+        }
+
+        let delete_key = SkipListKey::new(account, key.level, element);
+        let delete_trie_key = delete_key.key_hash();
+        let next_value = SkipListValue::<T>::read_trie(&delete_trie_key, trie)?;
+        trie.remove(delete_trie_key.as_slice())?;
+        if let Some(next_value) = next_value {
+            trie.insert(key.key_hash().as_slice(), &next_value.serialize())?;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn insert_item<T, D, R>(
+    account: AccountAddress,
+    element: &T,
+    trie: &mut EthTrie<D>,
+    rng: &mut R,
+) -> Result<(), TrieError>
+where
+    T: Clone + Listable + Ord + Serialize + DeserializeOwned + 'static,
+    D: DB,
+    R: Rng,
+{
+    let insert_key = SkipListKey::new(account, 0, element);
+
+    // The key is already present, so there is nothing to do.
+    if insert_key.trie_value(trie)?.is_some() {
+        return Ok(());
+    }
+
+    let head_key = SkipListHeadKey::<T>::new(account);
+    let trie_head_key = head_key.key_hash();
+    let head = SkipListHeadValue::<T>::read_trie(&trie_head_key, trie)?;
+    let end_of_list = SkipListValue::<T> { next_value: None }.serialize();
+
+    let Some(first_value) = head.first_value else {
+        // The list is empty, so we just insert this value
+        let new_head = SkipListHeadValue::new(0, element);
+        trie.insert(trie_head_key.as_slice(), &new_head.serialize())?;
+
+        trie.insert(insert_key.key_hash().as_slice(), &end_of_list)?;
+
+        return Ok(());
+    };
+
+    if element < first_value.as_ref() {
+        // The new element needs to become the head of the list.
+
+        // Randomly pick the height for the old head (now second element).
+        let insert_level = (pick_insert_height(rng) - 1) as u32;
```

Comment:
The `-1`'s purpose here evades me a bit. And also a bit of a nit:

```suggestion
        let reinsert_level = (pick_insert_height(rng) - 1) as u32;
```

as it's hard to keep in mind that it's meant for the *old* first node, not the new one.